crates/bevy_math/src/bezier.rs

```diff
@@ -0,0 +1,163 @@
+            + p[3] * t.powi(3)
+    }
+
+    /// Split the Bezier curve into `subdivisions`, and sample the position at each.
+    fn to_points(&self, subdivisions: i32) -> Vec<Self::Coord> {
+        (0..=subdivisions)
+            .map(|i| {
+                let t = i as f32 / subdivisions as f32;
+                self.evaluate_at(t)
+            })
+            .collect()
+    }
+}
+
+/// A 2-dimensional Bezier curve.
+#[derive(Default, Clone, Copy, Debug, PartialEq)]
+pub struct CubicBezier2d(pub [Vec2; 4]);
+
+impl CubicBezier for CubicBezier2d {
+    type Coord = Vec2;
+
+    fn control_points(&self) -> [Self::Coord; 4] {
+        self.0
+    }
+}
+
+impl CubicBezier2d {
+    /// Returns the point along the Bezier curve at the supplied parametric value `t`.
+    pub fn evaluate_at(&self, t: f32) -> Vec2 {
+        CubicBezier::evaluate_at(self, t)
+    }
+
+    /// Iterate over points in the bezier curve from `t = 0` to `t = 1`.
+    pub fn to_points(&self, subdivisions: i32) -> Vec<Vec2> {
+        CubicBezier::to_points(self, subdivisions)
+    }
+}
+
+/// A 3-dimensional Bezier curve.
+#[derive(Default, Clone, Copy, Debug, PartialEq)]
+pub struct CubicBezier3d(pub [Vec3; 4]);
+
+impl CubicBezier for CubicBezier3d {
+    type Coord = Vec3;
+
+    fn control_points(&self) -> [Self::Coord; 4] {
+        self.0
+    }
+}
+
+impl CubicBezier3d {
+    /// Returns the point along the Bezier curve at the supplied parametric value `t`.
+    pub fn evaluate_at(&self, t: f32) -> Vec3 {
+        CubicBezier::evaluate_at(self, t)
+    }
+
+    /// Iterate over points in the bezier curve from `t = 0` to `t = 1`.
+    pub fn to_points(&self, subdivisions: i32) -> Vec<Vec3> {
+        CubicBezier::to_points(self, subdivisions)
+    }
+}
+
+pub mod easing {
+    use super::CubicBezier;
+
+    /// Used to optimize cubic bezier easing, which only does operations in one dimension at a time,
+    /// and whose first and last control points are constrained to 0 and 1 respectively.
+    #[derive(Default, Clone, Copy, Debug, PartialEq)]
+    struct CubicBezier1d(f32, f32);
+
+    impl CubicBezier for CubicBezier1d {
+        type Coord = f32;
+
+        fn control_points(&self) -> [Self::Coord; 4] {
+            [0.0, self.0, self.1, 1.0]
+        }
+    }
+
+    /// A 2-dimensional Bezier curve used for easing in animation; the first and last control points
+    /// are constrained to (0, 0) and (1, 1) respectively.
+    #[derive(Default, Clone, Copy, Debug, PartialEq)]
+    pub struct CubicBezierEasing {
+        x: CubicBezier1d,
+        y: CubicBezier1d,
+    }
+
+    impl CubicBezierEasing {
+        /// Construct a cubic bezier curve for animation easing, with control points P1 and P2.
+        ///
+        /// This is equivalent to the syntax used to define cubic bezier easing functions in, say, CSS:
+        /// `ease-in-out = cubic-bezier(0.42, 0.0, 0.58, 1.0)`.
+        pub fn new(p1_x: f32, p1_y: f32, p2_x: f32, p2_y: f32) -> Self {
```

Comment:
should `p1` and `p2` be `Vec2`?

Reply:
In this case, no, I don't think so. Cubic Bezier easing is [commonly used](https://developer.mozilla.org/en-US/docs/Web/CSS/easing-function) with the syntax `cubic-bezier(0.42, 0.0, 1.0, 1.0)`. I've mirrored that syntax here as users will probably expect the constructor to look the same.

Reply:
If this were in `bevy_ui` I would agree, but not in `bevy_math`. CSS has one of the crappiest type system in the world in my opinion 😄

Reply:
I agree with Francois: more types please...

Reply:
> CSS has one of the crappiest type system in the world in my opinion 😄

FWIW, this is not just CSS, in fact I linked to the Android/Kotlin docs. This is also true for [flutter](https://api.flutter.dev/flutter/animation/Cubic/Cubic.html), [Qt](https://doc.qt.io/qt-6/qml-qtquick-pathcubic.html), [Elm](https://github.com/Dandandan/Easing/blob/master/Easing.elm#L161), and everywhere else I've looked. From what I can tell, this is basically standard, and would be a bit odd for a type designed for cubic Bezier animation not to have a constructor that takes an `f32` 4-tuple. I've also done a refactor pass based on other feedback that has resulted in the `CubicBezierEasing` type simply being a pair of `Vec2` control points.

Reply:
I found an implementation that is strongly typed, from Apple: https://developer.apple.com/documentation/uikit/uicubictimingparameters/1649854-init

Reply:
I think this is close enough to be the best of both worlds:

```rs
pub fn new(p1: impl Into<Vec2>, p2: impl Into<Vec2>)
```

Which means you can use it with minimal boilerplate:
```rs
let cubic_bezier = CubicBezierEasing::new((0.8, 0.0), (0.2, 1.0));
let t = cubic_bezier.ease(time);
```

Reply:
Good, I like that.